src/core/algorithms/dc/parser/dc_parser.cpp

```diff
@@ -0,0 +1,168 @@
+#include "algorithms/dc/parser/dc_parser.h"
+
+#include <stdexcept>
+
+#include <boost/algorithm/string.hpp>
+
+#include "algorithms/dc/model/dc.h"
+#include "algorithms/dc/model/operator.h"
+#include "algorithms/dc/model/predicate.h"
+#include "model/table/column_layout_relation_data.h"
+#include "model/table/typed_column_data.h"
+
+namespace mo = model;
+
+namespace {
+
+using namespace algos::dc;
+
+std::unique_ptr<ColumnOperand> GetOperand(ColumnOperand const& operand, std::string str_val) {
+    mo::Type const* type = operand.GetType();
+    return std::make_unique<ColumnOperand>(str_val, type);
+}
+
+}  // namespace
+
+namespace algos::dc {
+
+DCParser::DCParser(std::string dc_string, ColumnLayoutRelationData const* relation,
+                   std::vector<model::TypedColumnData> const& data)
+    : relation_(relation),
+      data_(data),
+      dc_string_(std::move(dc_string)),
+      has_next_predicate_(true),
+      cur_(0) {
+    str_operators_.reserve(str_operators_.size());
+    for (std::pair<frozen::string const, algos::dc::OperatorType> const& op_it :
+         Operator::kStringToOperatorType) {
+        std::string str_op = op_it.first.data();
+        str_operators_.emplace_back(std::move(str_op));
+    }
+};
+
+DC DCParser::Parse() {
+    boost::trim(dc_string_);
+
+    if (dc_string_.front() != '!') throw std::invalid_argument("Missing logic negation sign");
+    dc_string_.erase(dc_string_.begin());
+    boost::trim_left(dc_string_);
+
+    if (dc_string_.front() != '(' or dc_string_.back() != ')')
+        throw std::invalid_argument("Missing parenthesis");
+
+    dc_string_.erase(dc_string_.begin());
+    dc_string_.erase(std::prev(dc_string_.end()));
+
+    if (dc_string_.empty()) throw std::invalid_argument("Empty DC is not allowed");
+
+    std::vector<Predicate> res;
+    while (has_next_predicate_) {
+        Predicate pred = GetNextPredicate();
+        res.push_back(pred);
+    }
+
+    return {res};
+}
+
+Predicate DCParser::GetNextPredicate() {
+    size_t ind = dc_string_.find(sep_, cur_);
+    if (ind == std::string::npos) has_next_predicate_ = false;
+
+    std::string str_pred = dc_string_.substr(cur_, ind - cur_);
+    cur_ = ind + sep_.size();
+
+    return ConvertToPredicate(std::move(str_pred));
+}
+
+Predicate DCParser::ConvertToPredicate(std::string pred) {
+    size_t ind;
+    std::string str_operator;
+    for (std::string const& op : str_operators_) {
+        std::string spaced_op = ' ' + op + ' ';
+        ind = pred.find(spaced_op);
+        if (ind != std::string::npos) {
+            str_operator = op;
+            break;
+        }
+    }
+
+    if (ind == std::string::npos) throw std::invalid_argument("Missing permissible operator");
```

Comment:
This can be simplified:
```suggestion
auto some_name = [&pred](const std::string& op) {
    std::string spaced_op = ' ' + op + ' ';
    return pred.find(spaced_op) != std::string::npos;
};

auto it = std::ranges::find_if(str_operators_, some_name);
if (it == str_operators_.end()) {
    throw std::invalid_argument("Missing permissible operator");
}
```
we can extract the lambda to a variable, it will make for the